lib/spack/spack/install_test.py

```diff
@@ -52,6 +94,640 @@ def get_test_stage_dir():
+
+    Args:
+        pkg: package or package class of interest
+        add_virtuals: ``True`` adds test methods of provided package
+            virtual, ``False`` only returns test functions of the package
+
+    Returns:
+        names of non-empty test functions
+
+    Raises:
+        ValueError: occurs if pkg is not a package class
+    """
+    fns = test_functions(pkg, add_virtuals)
+    return [f"{cls_name}.{fn.__name__}" for (cls_name, fn) in fns]
+
+
+def test_functions(pkg: TestPackageType, add_virtuals: bool = False) -> List[Tuple[str, Callable]]:
+    """Grab all non-empty test functions.
+
+    Args:
+        pkg: package or package class of interest
+        add_virtuals: ``True`` adds test methods of provided package
+            virtual, ``False`` only returns test functions of the package
+
+    Returns:
+        list of non-empty test functions' (name, function)
+
+    Raises:
+        ValueError: occurs if pkg is not a package class
+    """
+    instance = isinstance(pkg, spack.package_base.PackageBase)
+    if not (instance or issubclass(pkg, spack.package_base.PackageBase)):  # type: ignore[arg-type]
+        raise ValueError(f"Expected a package (class), not {pkg} ({type(pkg)})")
+
+    pkg_cls = pkg.__class__ if instance else pkg
+    classes = [pkg_cls]
+    if add_virtuals:
+        vpkgs = virtuals(pkg)
+        for vname in vpkgs:
+            try:
+                classes.append((Spec(vname)).package_class)
+            except spack.repo.UnknownPackageError:
+                tty.debug(f"{vname}: virtual does not appear to have a package file")
+
+    # TODO (post-34236): Remove if removing empty test method check
+    def skip(line):
+        # This should match the lines in the deprecated test() method
+        ln = line.strip()
+        return ln.startswith("#") or ("warn" in ln and "deprecated" in ln)
+
+    doc_regex = r'\s+("""[\w\s\(\)\-\,\;\:]+""")'
+    tests = []
+    for clss in classes:
+        methods = inspect.getmembers(clss, predicate=lambda x: inspect.isfunction(x))
+        for name, test_fn in methods:
+            # TODO (post-34236): "test"->"test_" once remove deprecated methods
+            if not name.startswith("test"):
+                continue
+
+            # TODO (post-34236): Could remove empty method check once remove
+            # TODO (post-34236): deprecated methods though some use cases,
+            # TODO (post-34236): such as checking packages have actual, non-
+            # TODO (post-34236): empty tests, may want this check to remain.
+            source = re.sub(doc_regex, r"", inspect.getsource(test_fn)).splitlines()[1:]
+            lines = [ln.strip() for ln in source if not skip(ln)]
+            if not lines:
+                continue
+
+            tests.append((clss.__name__, test_fn))  # type: ignore[union-attr]
+
+    return tests
+
+
+def test_parts_process(pkg: Pb, test_specs: List[spack.spec.Spec], verbose: bool = False):
```

Comment:
(Refactor) If we could avoid having `test_*` functions in code outside of unit tests it would be great :sweat_smile: 
```suggestion
def process_test_parts(pkg: Pb, test_specs: List[spack.spec.Spec], verbose: bool = False):
```

Reply:
Fair enough.  I debated that myself though it is in `install_test` not `package_base`.

Changed.